[DOCS] Added documentation about --asm flag (#11212)
ethereum/solidity: docs/analysing-compilation-output.rst

.. index:: analyse, asm

#############################
Analysing the Compiler Output
#############################

> It is often useful to look at the assembly code generated by the compiler. However, the generated

Reviewer: Why "however"? This sounds like `--bin` would create "assembly code".

binary, i.e., the output of ``solc --bin contract.sol``, is difficult to read. It is recommended to use

> the flag ``--asm`` to analyse the assembly output.

Reviewer: Maybe add: Even for large contracts, looking a a graphical diff of the assembly before and after a change is often very enlightening.

Consider the following contract (named, say ``contract.sol``):

.. code-block:: Solidity

    // SPDX-License-Identifier: GPL-3.0

>     pragma solidity >= 0.8.0 < 0.9.0;

Reviewer: ```suggestion
    pragma solidity ^0.8.0;
```

Reviewer: Actually you can (and should) go back with the min version requirement a lot.

Author: Changed to 0.5.0

    contract C {
        function one() public pure returns (uint) {
            return 1;
        }
    }

The following would be the output of ``solc --asm contract.sol``

.. code-block:: none

>     ======= /tmp/test.sol:C =======

Reviewer: Can you compile it so that it just says `test.sol`?

          tag_10
          jump	// in
        tag_15:
            /* "#utility.yul":229:353   */
          swap3
          swap2
          pop
          pop
          jump	// out
            /* "#utility.yul":359:436   */
        tag_13:
          0x00
            /* "#utility.yul":425:430   */
          dup2
            /* "#utility.yul":414:430   */
          swap1
          pop
            /* "#utility.yul":404:436   */
          swap2
          swap1
          pop
          jump	// out

        auxdata: 0xa2646970667358221220847c9788d95894e484e81d4710aff9be15a68f619a87830acd37d4ad818af99764736f6c63782b302e382e342d646576656c6f702e323032312e332e32362b636f6d6d69742e64373164313530352e6d6f64005c
   }

Alternatively, the above output can also be obtained from `Remix <https://remix.ethereum.org/>`_,
under the option "Compilation Details" after compiling a contract.

Notice that the ``asm`` output starts with the creation / constructor code. The deploy code is
provided as part of the sub object (in the above example, it is part of the sub-object ``sub_0``).
The ``auxdata`` field corresponds to the contract :ref:`metadata
<encoding-of-the-metadata-hash-in-the-bytecode>`.

> 

Reviewer: Can you explain that the comments tell you which source part the code was generated for and that `#utility.yul` is an internally-generated file of utility functions that can be obtained using `--combined-json generated-sources,generated-sources-runtime`?

Author: Added this.

Similarly, the optimized assembly can be obtained with the command: ``solc --optimize --asm

> contract.sol``. Often times, it is interesting to see if two different code in Solidity results in

Reviewer: ```suggestion
contract.sol``. Often times, it is interesting to see if two different sources in Solidity result in
```

.. index:: analyse, asm

#############################

> Analysing the Compiler Output

Reviewer 2: I'm always confused if we should use the American or British spelling `analyzing` vs `analysing` :-)

Author: I think docs use British spelling. So `analyse` should work, right?

#############################

It is often useful to look at the assembly code generated by the compiler. The generated binary,
i.e., the output of ``solc --bin contract.sol``, is generally difficult to read. It is recommended
to use the flag ``--asm`` to analyse the assembly output. Even for large contracts, looking at a

> graphical diff of the assembly before and after a change is often very enlightening.

Reviewer 2: `visual` instead of `graphical` diff?

          dup5
            /* "#utility.yul":275:346   */
          tag_10
          jump	// in
        tag_15:
            /* "#utility.yul":229:353   */
          swap3
          swap2
          pop
          pop
          jump	// out
            /* "#utility.yul":359:436   */
        tag_13:
          0x00
            /* "#utility.yul":425:430   */
          dup2
            /* "#utility.yul":414:430   */
          swap1
          pop
            /* "#utility.yul":404:436   */
          swap2
          swap1
          pop
          jump	// out

        auxdata: 0xa2646970667358221220a5874f19737ddd4c5d77ace1619e5160c67b3d4bedac75fce908fed32d98899864736f6c637827302e382e342d646576656c6f702e323032312e332e33302b636f6d6d69742e65613065363933380058
    }

Alternatively, the above output can also be obtained from `Remix <https://remix.ethereum.org/>`_,
under the option "Compilation Details" after compiling a contract.

Notice that the ``asm`` output starts with the creation / constructor code. The deploy code is
provided as part of the sub object (in the above example, it is part of the sub-object ``sub_0``).
The ``auxdata`` field corresponds to the contract :ref:`metadata

> <encoding-of-the-metadata-hash-in-the-bytecode>`. The comments in the assembly output points to the

Reviewer 2: ```suggestion
<encoding-of-the-metadata-hash-in-the-bytecode>`. The comments in the assembly output point to the
```

Author: fixed both the comments.

source location. Note that ``#utility.yul`` is an internally generated file of utility functions
that can be obtained using the flags ``--combined-json
generated-sources,generated-sources-runtime``.

Similarly, the optimized assembly can be obtained with the command: ``solc --optimize --asm
contract.sol``. Often times, it is interesting to see if two different sources in Solidity result in
the same optimized code. For example, to see if the expressions ``(a * b) / c``, ``a * b / c``
generates the same bytecode. This can be easily done by taking a ``diff`` of the corresponding

> assembly output.

Reviewer: Potentially after stripping the comments that reference the source location.

> 

Author: Added this line.
